Fix for launching Brackets from command line (Win Only)

nethip opened the pull request:
The change basically fixes the following problems. 

1) Our command line arguments were never respected as Brackets was expecting a full path and we were just passing the file names

2) Adding the install location to the PATH variable which will allow users to launch Brackets just by typing "Brackets" instead of full path like "C:\Program Files (86)\Brackets\Brackets.exe".

This is a Win only change.

nethip commented:
@JeffryBooher @ingorichter Please review with when you guys get some time.

nethip commented:
Adding @RaymondLim @redmunds

abose commented on appshell/cefclient_win.cpp:
```diff
@@ -188,7 +245,7 @@ int APIENTRY wWinMain(HINSTANCE hInstance,
   else {
 	// If the shift key is not pressed, look for the index.html file 
-	if (GetAsyncKeyState(VK_SHIFT) == 0) {
+	if (!isShiftKeyDown) {
```
Shift-click to launch brackets from source verified?

nethip replied:
This checked is moved to the top. And the check again is replaced with checking for a bit set. When launching from command line, `GetAsyncKeyState(VK_SHIFT) == 0` was returning true. The correct fix is to check for the set bit `GetAsyncKeyState(VK_SHIFT) & 0x8000`.

abose commented on installer/win/Brackets.wxs:
```diff
@@ -96,7 +104,19 @@
-                <Directory Id='INSTALLDIR' Name='!(loc.ShortProductName)'/>
+                <Directory Id='INSTALLDIR' Name='!(loc.ShortProductName)'>
+                  <Component Id="UpdatePath"
+                             Guid="1781A625-8ACB-45E7-A8BA-219D81760B2E">
```
What is GUID here?

nethip replied:
This is mandated by WIX.

abose commented on installer/win/Brackets.wxs:
```diff
@@ -96,7 +104,19 @@
+                             Guid="1781A625-8ACB-45E7-A8BA-219D81760B2E">
+                    <CreateFolder />
+                    <Environment Id="PATH"
+                                 Action="set"
```
what does this section do.

nethip replied:
This appends `[INSTALLDIR]\Brackets.exe` to the PATH environment variable making Brackets launchable  from command line.

abose commented on installer/win/Brackets_fr-fr.wxl:
```diff
@@ -53,7 +53,7 @@
     <String Id="VerifyReadyDlgInstallText" Overridable="yes">Cliquez sur Installer pour lancer la procédure ou sur Annuler pour quitter.</String>
     <String Id="VerifyReadyDlgRemoveText" Overridable="yes">Cliquez sur Supprimer pour désinstaller !(loc.ShortProductName) ou sur Annuler pour quitter.</String>
-    
+    <String Id="OpenWithText" Overridable="yes">Open With !(loc.ShortProductName)</String>
```
WHy only fr-fr?

nethip replied:
We only have two language installers. And this is just a placeholder before our LOC team subsitutes with french translation.

le717 commented:
No no no no. I object to adding Brackets to the PATH. That is usually reserved for runtime utilities or interpreted programming languages (python, java, node) or critical items needed for system stability (PowerShell, GPU drivers, system32). I don't think we need to add Brackets to the PATH just to get command-line arguments working. There's another way to do it. Let me find it real quick.

le717 commented:
Found it. https://github.com/adobe/brackets/issues/7482#issuecomment-40162534 ([referenced section](https://msdn.microsoft.com/en-us/library/windows/desktop/ee872121%28v=vs.85%29.aspx#reg_app)). I think that should work in the same manner as adding it to the PATH, but with less risk. FWIW, adding applications to the system PATH is generally frowned upon, and when the implementation is flawed the entire PATH can be wrecked (I've had this happen before). I highly advise a followup PR to replace the PATH editing with this registry entry (after confirming it works, which I am pretty sure it does) in this same release (1.3).

nethip commented:
@le717 Thanks for sharing the link. If we could do this with a registry entry we could very well do that easily. Let me go through the link and consider other alternatives.

nethip commented:
@le717 The registry entry only takes care of launching Brackets using ***Run Command...*** not through command line. To be more technical, any tool that uses `ShellExecuteEx()` to launch a new process will be able to successfully find the target app with an entry in *HKLM\SOFTWARE\Microsoft\Windows\AppPaths*. But I doubt if **cmd.exe** or **GitBash** use `ShellExecuteEx()` to launch apps.

Once thing to note here is that we are relying on the MSI that is the standard windows installer that for sure understands how to manage the PATH update efficiently (I have tested this with variety of test cases and the installer does a good job of undoing whatever has been done through the installation process, during uninstallation). It would have been a bigger problem if we were updating the PATH variable programatically. 

 Another option is to create an alias to Brackets, in Windows/System32 folder, which is even more uglier than updating the PATH variable. 

Considering the no of votes **81** for this trello card (https://trello.com/c/dRNqhd2L/349-s-open-file-folder-from-command-line-on-win-and-linux) I think we should support this feature.

le717 commented:
> Considering the no of votes for this trello card I think we should support this feature.

I agree this feature should be added, but I don't agree this is _how_ it should be added.

> The registry entry only takes care of launching Brackets using Run Command... not through command line. To be more technical, any tool that uses ShellExecuteEx() to launch a new process will be able to successfully find the target app with an entry in HKLM\SOFTWARE\Microsoft\Windows\AppPaths. But I doubt if cmd.exe or GitBash use ShellExecuteEx() to launch apps.

The run command is a command prompt, but only runs one command rather than many. Also, did you compare the .bat script with the MSDN docs on registering an application? I think this method will work. I have Word installed and I can open it from the command-line, and it is not in my PATH nor is there a symlink in System32 (which agreed is ugly).

nethip commented:
I am open towards making this change if this can be done through a registry entry. I could not find a way of to do this. In fact I would recommend this registry way. Can you share the details of .bat registration in the registry. It would surely help me in doing this change.

le717 commented:
OK, I apologize. I have always assumed the run command was a mini-command prompt. I just tested it now, and you are right, it only works for Run. I still think it is possible without the PATH...

le717 commented:
Just so I'm totally clear, adding this to the PATH permits what, again? Opening files/folders in Brackets using open with, or opening a command prompt and typing `brackets "some/file/here.js"`?

nethip commented:
No worries. Let me know if you if you come across any interesting solitons to fix this. 

To answer to your last question it does both. We have changed the installer to add open with menu to file and folder context menus. Also Brackets can be launched from command line with file/folder name as argument.

le717 commented:
I found this a while back. It uses the registry to open files from the context menus. https://gist.github.com/wormeyman/75e3036993136c8b1830

As for the command line, I know of 0 programs that add themselves to the PATH just to open files like that. As I said, it is kinda frowned upon to add a program to the PATH in general unless it is an absolute last resort to do something or it is a purely command line utility (git, python, java, etc). GUI programs like Brackets do not fit into those use cases. In fact, adding to the PATH works around functionality Windows already has. I found that after adding that app registration registry key, I could run `start brackets "path\to\file.js"` and `start brackets "path\to\folder` and it worked. The `start` command is an underused and underexposed gem, and that is what should be used in this case, not adding Brackets (a GUI'd code IDE) to the PATH.

Maybe a forum topic should be raised asking about this issue. As I said, I know of zero IDEs that add themselves to the PATH, and it almost sounds to me this is a case of "The user doesn't quite know what they are asking". Also, the 81 votes were probably more towards the context menus than adding to the PATH.

peterflynn commented on appshell/cefclient_win.cpp:
```diff
@@ -140,28 +194,31 @@ int APIENTRY wWinMain(HINSTANCE hInstance,
+        return 0;
+      }
+    }
+    // otherwise, fall thru and launch a new instance
```
Belated comment: all the indent here looks off.  Did this PR change spaces to tabs, or vice versa?

nethip replied:
I have changed all the indentation from tab to space.

peterflynn commented:
@abose @nethip Let's discuss this a little more before committing to ship as-is -- given that this is optional or not even a feature in many (most?) other Windows editors

peterflynn commented:
@le717 It's a popular feature in Mac code editors to be able to launch from the command line -- e.g. `mate <filename>` to launch TextMate.  Presumably the use cases would be equally as common on Windows.  And that kind of shorthand _only_ works if an app is in the PATH.  That said, in most other apps the user has to either manually modify their PATH, or there's a checkbox in the installer for opting in/out.  That seems like the safer bet for Brackets too...

peterflynn commented on appshell/cefclient_win.cpp:
```diff
@@ -83,30 +83,82 @@ bool IsFilename(const std::wstring& str) {
+
+  DWORD retval;
+  TCHAR  buffer[MAX_UNC_PATH] = TEXT(""); 
+  TCHAR  buf[MAX_UNC_PATH]    = TEXT(""); 
```
Belated nit: looks like this 2nd buffer variable is unused.

nethip replied:
Will remove this.

ryanstewart commented:
Sorry for jumping in late. Just did a bit of research, and like @peterflynn I think there's some confusion/difference between the Windows and Mac experience. On Mac, as Peter says, this is a pretty common feature. TextMate and Sublime both enable it. I assumed Sublime behaved the same way on Windows, but it looks like that's not the case. Googling around it looks like you *can* do this with Sublime, but it's a [very manual process](https://github.com/adobe/brackets-shell/pull/510#issuecomment-91660156).

I think most of the votes in the card referenced are from Mac users who want to be able to use Brackets they way they use TextMate or Sublime. 

So the big question is whether or not we want to *automatically* do that for Windows users. I kind of think putting it in the debug menu is enough but maybe we need a more explicit warning. @le717 do you think this would be worth emailing out to the Brackets-Dev group about and getting feedback?

peterflynn commented:
@ryanstewart Even on Mac it's not that clear cut -- I'm pretty sure TextMate requires you to manually modify your path or add an alias somewhere.  Not sure if any editors other than Sublime automatically add themselves.

Re votes on the card: until this week I think the title/description was Mac-<i>only</i>, so that is probably a safe assumption about what the voters wanted :-)

le717 commented:
>So the big question is whether or not we want to automatically do that for Windows users. I kind of think putting it in the debug menu is enough but maybe we need a more explicit warning. @le717 do you think this would be worth emailing out to the Brackets-Dev group about and getting feedback? 

Yes, I do think this is worth getting feedback from Brackets-Dev.

nethip commented:
@peterflynn @ryanstewart I am not sure if only MAC users had gone and up voted this considering the fact that there are multiple mentions of this being supported on Windows.

* * Most of the required work is covered by #280 (_superseded by https://trello.com/c/U7pQ0yIg/117-5-open-with-brackets-option-from-windows-explorer_), e.g. parsing command line args as things to open, and making sure we reuse an existing instance of Brackets if open
* *  Add our executable to the PATH at install time.  See RE user story: https://trello.com/c/NRdWorbu/32-br-co-add-to-path
* *  Ensure Windows matches this behavior too

We could do what Peter had suggested on Windows like providing a checkbox for users to opt out of this.

MarcelGerber commented:
I also think this shouldn't be the default behavior on Windows. As @le717 mentioned, usually only command-line only tools add themselves to the PATH; I've got Sublime Text 3 and Notepad++ installed over here and neither of them added themselves to PATH.
Also, currently, it could happen that Brackets node.exe (in the install directory) is called instead of the actual node.exe if Brackets' PATH entry comes before the other one. That would be very, very, very bad.

nethip commented:
@MarcelGerber That's a very good catch. We are thinking of making this PATH change optional in the installer, through a check box. Now I am thinking should we even do that. 

However, we are going to add one checkbox in the installer for adding "open with Brackets" menu to file and explorer menu.

le717 commented:
@MarcelGerber brought up a _very_ good point with node, and such a problem can happen very easily  (just install Node and let the installer add it to the PATH, then use this PR to install Brackets. Tada!). As I've stated, adding Brackets to the PATH does not need to happen at all. If people want to load files from the command-line, registering Brackets and using `start` (as I stated in https://github.com/adobe/brackets-shell/pull/510#issuecomment-91660156) will do.

> However, we are going to add one checkbox in the installer for adding "open with Brackets" menu to file and explorer menu. 

Yes, please do. That's been a long time complaint about the Windows installer (see the very long discussion and resulting Trello card at adobe/brackets#6073), and making this an (I might suggest default, but we'll see) option will help remedy those complaints.

peterflynn commented:
@MarcelGerber Very good catch!  Too bad we haven't yet figured out how to rename the executable back to "Brackets-node.exe," since that would have mitigated the problem...

@nethip Seems like we could solve using a script for indirection, similar to what we do on Mac.  E.g. add a "brackets.bat" to the build in its own directory (`INSTALL_DIR\launcher\brackets.bat`?), and then add that dir to the path instead of the root install dir.

nethip commented:
@peterflynn @MarcelGerber @le717 @ryanstewart 

I have created a new PR for this https://github.com/adobe/brackets-shell/pull/514.

These are the changes in nutshell
* Made the update path and context menu addition to file and folder context menus optional.
* The PATH varible will now be [INSTALLDIR]/command.
* Changed project setup scripts to create commands/brackets.bat by copying it from scripts/brackets.bat
* Changed the Gruntfile to copy the new folder command to Brackets staging area.

So basically now we have a new folder called [INSTALLDIR]\command which contains the script brackets.bat. This is what is going to be added to the PATH so that anyone when type `brackets` from command line, it is going to invoke this script. Basically this script will contain this.

```
@echo off
start Brackets.exe $*
```
`start` command now will be able to find Brackets.exe as we now have a new entry in Registry which will make Brackets runnable from `Run Command` as well.

`HKLM\SOFTWARE\Microsoft\Windows\CurrentVersoon\App Paths\Brackets.exe`

I have put in some placeholder string for the checkboxes inside the installer. Feel free to suggest strings for these.

* Update PATH environment variable with Brackets install location ( in order to launch Brackets from Command Line)
* Add "Open With Brackets" menu to file and folder context menus
